src/test/ui/borrowck/index-mut-help.nll.stderr

```diff
@@ -0,0 +1,22 @@
+error[E0596]: cannot borrow data in a `&` reference as mutable
+  --> $DIR/index-mut-help.rs:21:5
+   |
+LL |     map["peter"].clear();           //~ ERROR
+   |     ^^^^^^^^^^^^ cannot borrow as mutable
```

Comment:
Can you also add the message to the NLL errors? Eventually we'll switch over and lose the new note otherwise.

Reply:
I tried to do that in the past couple of hours and now I really don't know how to proceed. 

So I'm pretty sure I have to add my code to `report_mutability_error()` in `librustc_mir/borrow_check/mutability_errors.rs`. But I don't know how I can get the information whether this error was caused by an index operation. I have access to the following things:

- `access_place` and `the_place_err`: these were my biggest hope of getting information about the expression causing the error. Sadly, in all three cases mentioned in this thread, they are `ProjectionElem::Deref { base: Place::Local }` which doesn't help me. I would have expected one of them to be a `ProjectionElem::Index`.
- `span` and `error_access`: not helpful AFAICT
- `location`: it's only a reference to the enclosing basic block, right? So not really helpful either
- `self.mir_def_id`: this is a `DefId` and thus references a definition of an item, right? So I guess it references the enclosing function. So not useful either.
- `self.{other_fields}`: AFAICT most of them are global structures and nothing pointing to the expression in question. I don't think anything is helpful here.

So I'm stuck and can't invest a lot more time into this. Could someone help me to get this done fairly quickly? If you people in this thread don't know about this stuff, could you ping someone who does know the code? Or tell me where I should go and ask about this?

Reply:
`ProjectionElem::Index` is only for simple arrays and slices. Everything else goes through the `IndexMut`/`Index` trait methods. I'm assuming rustc decides on `Index` due to the lack of `IndexMut` and then tries to take a mutable reference to the result of the method call (which is where your `Deref` comes from: `&mut Index::index(&map, "peter")`)

cc @rust-lang/wg-compiler-nll any ideas how to do this in the MIR borrowchecker? Right now rustc (with the nll feature gate active) produces

```
error[E0596]: cannot borrow data in a `&` reference as mutable
 --> src/main.rs:7:5
  |
7 |     map["peter"].clear();           //~ ERROR
  |     ^^^^^^^^^^^^ cannot borrow as mutable
```

which actually is less informative than the message of old borrowck even before this PR: `cannot borrow immutable indexed content as mutable`

Reply:
I think the thing to do would be to look at the `&` reference and try to figure out where it came from. It'll probably be a temporary here, which is a good indicator that we should try to get more info about it. We'd see it as the result of an `Index` call -- that is probably a good signal for us to insert this advice, since unless the call to `Index` was manually written by the user, it must imply that `IndexMut` does not exist.

Reply:
maybe worth filing a bug for it?

Reply:
Thanks for your advice! Would it be possible to create a bug about this and then merge this PR as is? I don't think I will have time to work on this anytime soon. Then someone else can work on adding the same advice to the NLL version.

Reply:
Filed #53228 to follow up.

Reply:
Thanks :)